src/model/chart-model.ts

```diff
@@ -467,22 +467,30 @@ export class ChartModel implements IDestroyable {
 	}
 
 	public updateTimeScale(newBaseIndex: TimePointIndex, newPoints?: readonly TimeScalePoint[]): void {
+		const oldFirstTime = this._timeScale.indexToTime(0 as TimePointIndex);
+
 		if (newPoints !== undefined) {
 			this._timeScale.update(newPoints);
 		}
 
+		const newFirstTime = this._timeScale.indexToTime(0 as TimePointIndex);
+
 		const currentBaseIndex = this._timeScale.baseIndex();
 		const visibleBars = this._timeScale.visibleStrictRange();
 
 		// if time scale cannot return current visible bars range (e.g. time scale has zero-width)
 		// then we do not need to update right offset to shift visible bars range to have the same right offset as we have before new bar
 		// (and actually we cannot)
-		if (visibleBars !== null) {
+		if (visibleBars !== null && oldFirstTime !== null && newFirstTime !== null) {
 			const isLastSeriesBarVisible = visibleBars.contains(currentBaseIndex);
+			const isFirstSeriesBarVisible = visibleBars.contains(0 as TimePointIndex);
+			const isFirstTimeChanged = oldFirstTime.timestamp > newFirstTime.timestamp;
+			const isDataAdded = newBaseIndex > currentBaseIndex;
+			const isDataAddedToRight = isDataAdded && !isFirstTimeChanged;
+			const isVisibleRangeFilled = !isLastSeriesBarVisible && !isFirstSeriesBarVisible;
```

Comment:
Maybe we need to call it something like `areDataEndsVisible`? @Nipheris can you please take a look here?

Reply:
`areDataEndsVisible` means like letf end of end and right end of data visible inside current timescale? 
```suggestion
			const areDataEndsVisible = isLastSeriesBarVisible && isFirstSeriesBarVisible;
```
But we need shift timescale when all ends of data out of timescale and bars added to right side of data.
Expression like this `if (isDataAddedToRight && !areDataEndsVisible )` will perform the shift when the right edge is visible, which we don't need.

May be `areDataEndsInvisible`?

Reply:
> May be areDataEndsInvisible?

Yeah, I think it was typo, I'm sorry 😂 In terms of "data ends are visible" I guess we need to add _any_ there (`IsAnyDataEndVisible`) and make check "negative"